### DISCUSSION TITLE
Couldn't determine job existence

### DISCUSSION THREAD

**Discussion by viveknuna** [quartznet/quartznet#2149] [Q&A]
We are intermittently getting the Couldn't determine job existence exception while running the Immediate job. I can't precisely reproduce it.

It occurs when calling [this ](https://github.com/quartznet/quartznet/blob/main/src/Quartz/IScheduler.cs#L313)method.

I am passing whenOptions as an empty string.

```
var triggerBuilder = TriggerBuilder.Create().WithIdentity(triggerName, triggerGroup);
var result = ExtendTriggerBuilder(ref triggerBuilder, whenOptions);
```
```
  job = JobBuilder.Create<JobType>()
      .WithIdentity(JobName, JobGroup)
      .UsingJobData("jobType", GetJobType())
      .UsingJobData("tenantId", TenantManager.TenantContext.TenantId.ToString())
      .UsingJobData("tenantDbConfig", dbConfig)
      .UsingJobData("serviceId", serviceId)
      .UsingJobData("whenOptions", whenOptions)
      .UsingJobData("correlationId", TenantManager.TenantContext.CorrelationId?.ToString() ?? Guid.Empty.ToString())
      .UsingJobData("payload", payload)
      .Build();
```
`var result = await scheduler.ScheduleJob(job, trigger);`

**Trigger and Job format:**

```
protected string TriggerGroup => $"trigger_{JobTypeString}_{TenantManager.TenantContext.TenantId}";

/// <summary>
///     Trigger Name.
/// </summary>
protected string TriggerName => $"trigger_{TenantManager.TenantContext.CorrelationId}_{TenantManager.TenantContext.MessageId}";

/// <summary>
///     Job group Name.
/// </summary>
protected string JobGroup => $"schedule_generic_job_{JobTypeString}_{TenantManager.TenantContext.TenantId}";

/// <summary>
///     Job Name.
/// </summary>
protected string JobName => $"job_{TenantManager.TenantContext.CorrelationId}_{TenantManager.TenantContext.MessageId}";
```

**Exception Details:**

**Message**: Couldn't determine job existence (schedule_generic_job_immediate_busmessagejob_11111111-1111-1111-1111-84ba20252626.job_46ee9029-9f58-4b37-b3cd-475009073171_70f23160-6818-4864-8d04-755325f460eb): Execution Timeout Expired.  The timeout period elapsed prior to completion of the operation or the server is not responding.

**problemId**: System.ComponentModel.Win32Exception at Quartz.Impl.AdoJobStore.JobStoreSupport+<JobExists>d__161.MoveNext
**type**: System.ComponentModel.Win32Exception
**Inner Exception**: Execution Timeout Expired.  The timeout period elapsed prior to completion of the operation or the server is not responding. / Unknown error 258
**trace**:at Quartz.Impl.AdoJobStore.JobStoreSupport.JobExists(ConnectionAndTransactionHolder conn, JobKey jobKey, CancellationToken cancellationToken)
   at Quartz.Impl.AdoJobStore.JobStoreSupport.StoreJob(ConnectionAndTransactionHolder conn, IJobDetail newJob, Boolean replaceExisting, CancellationToken cancellationToken)
   at Quartz.Impl.AdoJobStore.JobStoreSupport.<>c__DisplayClass156_0.<<StoreJobAndTrigger>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at Quartz.Impl.AdoJobStore.JobStoreSupport.ExecuteInNonManagedTXLock[T](String lockName, Func`2 txCallback, Func`3 txValidator, Nullable`1 requestorId, CancellationToken cancellationToken)
   at Quartz.Impl.AdoJobStore.JobStoreSupport.ExecuteInNonManagedTXLock[T](String lockName, Func`2 txCallback, Func`3 txValidator, Nullable`1 requestorId, CancellationToken cancellationToken)
   at Quartz.Impl.AdoJobStore.JobStoreSupport.StoreJobAndTrigger(IJobDetail newJob, IOperableTrigger newTrigger, CancellationToken cancellationToken)
   at Quartz.Core.QuartzScheduler.ScheduleJob(IJobDetail jobDetail, ITrigger trigger, CancellationToken cancellationToken)
   at Automation.Management.Adapters.Repositories.GenericScheduler.GenericScheduler`1.Schedule(String whenOptions, String payload, String serviceId) in /src/Management/Automation.Management.Adapters/Repositories/GenericScheduler/GenericScheduler.cs:line 109
   at Automation.Management.Adapters.Repositories.BusMessageSchedulerRepository.Schedule(ScheduleType scheduleType, String whenOptions, String payload, String serviceId) in /src/Management/Automation.Management.Adapters/Repositories/BusMessageSchedulerRepository.cs:line 49
   at Automation.Management.Quartz.Service.Controllers.SchedulerController.Generic(ScheduleType when, String serviceId, String payload, String whenOptions) in /src/Management/Automation.Management.Quartz.Service/Controllers/SchedulerController.cs:line 81
**Quartz version: 3.5.0
.Net framework: .Net 6**

Below are the quartz-related configuration details.

```
["quartz.scheduler.instanceName"] = $"QuartzScheduler_{tenantId}_{tenantDbConfigHash}",
["quartz.scheduler.instanceId"] = "AUTO",

// Is the amount of time in milliseconds that the scheduler will wait before re-queries for available triggers when the scheduler is otherwise idle.
["quartz.scheduler.idleWaitTime"] = "300000", // default: 30000

// You normally want to use JobStoreTX unless unless you want to integrate to some transaction-aware framework
["quartz.jobStore.type"] = "Quartz.Impl.AdoJobStore.JobStoreTX, Quartz",
// optional: if "true", can avoid versioning issues seen with BLOB data
["quartz.jobStore.useProperties"] = "true", // default: false
// required: this must match the value used in the "quartz.dataSource.*" properties
["quartz.jobStore.dataSource"] = $"{tenantId}",
// optional: the number of milliseconds the scheduler will 'tolerate' a trigger to pass its next-fire-time by
["quartz.jobStore.misfireThreshold"] = "60000", // default: 60000

// optional: frequency (in milliseconds) at which this instance "checks-in" with the other instances of the cluster
["quartz.jobStore.clusterCheckinInterval"] = "15000", // default: 15000
// optional: required if there are multiple instances of quartz using same tables
["quartz.jobStore.clustered"] = "false", // default: false; should be false if only 1 pod
// required: if set correctly, this should auto-configure many of the properties
["quartz.jobStore.driverDelegateType"] = "Quartz.Impl.AdoJobStore.SqlServerDelegate, Quartz",
// optional: required for us, since we use customized prefixes
["quartz.jobStore.tablePrefix"] = UnoConstants.DatabaseAutomationSchema + ".QRTZ_",

// Required: a datasource (and serializer type) must be specified
["quartz.serializer.type"] = "json",
[$"quartz.dataSource.{tenantId}.connectionString"] = tenantDbConfig,
[$"quartz.dataSource.{tenantId}.validationQuery"] = "SELECT 1",
[$"quartz.dataSource.{tenantId}.provider"] = "SqlServer",

[$"quartz.scheduler.typeLoadHelper.type"] = "Automation.Management.Adapters.Quartz.QuartzTypeLoadHelper, Automation.Management.Adapters",

// optional: should not be necessary if "quartz.jobStore.driverDelegateType" is correct.
// The below are to fix "Failure obtaining db row lock" error. See: https://github.com/quartz-scheduler/quartz/issues/268
//[$"quartz.jobStore.lockHandler.type"] = "Quartz.Impl.AdoJobStore.StdRowLockSemaphore",//UpdateLockRowSemaphore,StdRowLockSemaphore - this should not be needed as Quartz will pick the best lock type.

// maxRetry and retryPeriod are not applicable to all types of lock handler (e.g. UpdateLockRowSemaphore).
// Unsupported lock handlers may trigger an exception.
// optional: do not use if "quartz.jobStore.lockHandler.type" is "Quartz.Impl.AdoJobStore.UpdateLockRowSemaphore"
//[$"quartz.jobStore.lockHandler.maxRetry"] = "7", // Default 3
// optional: do not use if "quartz.jobStore.lockHandler.type" is "Quartz.Impl.AdoJobStore.UpdateLockRowSemaphore"
//[$"quartz.jobStore.lockHandler.retryPeriod"] = "3000", // Default 1000

// optional: default should be sufficient for nearly every user
//[$"quartz.threadPool.type"] = "Quartz.Simpl.SimpleThreadPool, Quartz", // default: "Quartz.Simpl.DefaultThreadPool"

// number of threads available to execute jobs
// bigger numbers allow more jobs to fire simultaneously, but at expense of system resources
// suggested value = between 1-100; less than 5 should be sufficient for most cases
// rule of thumb: use up to 3 x the number of processors
[$"quartz.threadPool.threadCount"] = "18", // default: -1, which equates to 10
// max number of triggers that can be batched at once
// bigger numbers = more efficient, but bigger chance of improper balancing between nodes
// suggested value = same as number of executor threads
[$"quartz.scheduler.batchTriggerAcquisitionMaxCount"] = "18", // default: 1
// time window over which triggers are batched (milliseconds)
// bigger numbers = more likely batch acquistion of triggers, but at cost of precision of firing time 
//[$"quartz.scheduler.batchTriggerAcquisitionFireAheadTimeWindow"] = "1000" // default: 0
```

I also checked there was not much load on the database when the exception occurred.

**Comment by viveknuna** [comment]
@lahma Do you have any suggestions?

**Comment by lahma** [comment]
Could be many things, like

* missing indexes in database
* backups running at the given moment
* noisy other tables that cause load
* querying/modifying the tables directly without using the `IScheduler` API

**Reply by viveknuna** [reply]
@lahma I have checked all 4 points mentioned by you. I have checked, it looks like issue is something else.

can I share my source code with you? Will you be able to help me?
